LLC purview Account

chunyu3 opened the pull request:
# All SDK Contribution checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.
- [ ] **Please open PR in `Draft` mode if it is:**
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md).**
- [ ] **The pull request does not introduce [breaking changes](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md).**

### [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#general-guidelines)
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code. (Track 2 only)
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK. Please double check nuget.org current release version.

## Additional management plane SDK specific contribution checklist: 
Note: Only applies to `Microsoft.Azure.Management.[RP]` or `Azure.ResourceManager.[RP]`
 
- [ ] Include updated [management metadata](https://github.com/Azure/azure-sdk-for-net/tree/main/eng/mgmt/mgmtmetadata).
- [ ] Update AzureRP.props to add/remove version info to maintain up to date API versions.

### Management plane SDK Troubleshooting
- If this is very first SDK for a services and you are adding new service folders directly under /SDK, please add `new service` label and/or contact assigned reviewer.
- If the check fails at the `Verify Code Generation` step, please ensure:
	- Do not modify any code in generated folders.
	- Do not selectively include/remove generated files in the PR.
	- Do use `generate.ps1/cmd` to generate this PR instead of calling `autorest` directly.
	Please pay attention to the @microsoft.csharp version output after running `generate.ps1`. If it is lower than current released version (2.3.82), please run it again as it should pull down the latest version.
	
	**Note: We have recently updated the PSH module called by `generate.ps1` to emit additional data. This would help reduce/eliminate the Code Verification check error. Please run following command**:

	    `dotnet msbuild eng/mgmt.proj /t:Util /p:UtilityName=InstallPsModules`

### Old outstanding PR cleanup
 Please note:
	If PRs (including draft) has been out for more than 60 days and there are no responses from our query or followups, they will be closed to maintain a concise list for our reviewers.

chunyu3 committed:
LLC purview Account

chunyu3 committed:
use default credential

ellismg commented on sdk/purview/Azure.Analytics.Purview.Account/README.md:
````diff
@@ -0,0 +1,114 @@
+Install the Azure Purview Account client library for .NET with [NuGet][client_nuget_package]:
+
+```
+dotnet add package Azure.Analysis.Purview.Account --prerelease
````
Should be `dotnet add dotnet add package Azure.Analytics.Purview.Account`

ellismg commented on sdk/purview/Azure.Analytics.Purview.Account/CHANGELOG.md:
```diff
@@ -0,0 +1,11 @@
+## 1.0.0-beta.2 (Unreleased)
+
+
+## 1.0.0-beta.1 (2021-05-12)
```
Change the date back to (Unreleased) and drop the 1.0.0-beta.2 section.

ellismg commented on sdk/purview/Azure.Analytics.Purview.Account/api/Azure.Analytics.Purview.Account.netstandard2.0.cs:
```diff
@@ -0,0 +1,213 @@
+namespace Azure.Analytics.Purview.Catalog
```
This doesn't seem like the API surface for the library.  Can you please run the `Export-Api.ps1` script to update this file (it will also fix a leg in CI)

ellismg commented on sdk/purview/Azure.Analytics.Purview.Account/src/Generated/AccountsClient.cs:
```diff
@@ -0,0 +1,507 @@
+        /// <param name="credential"> A credential used to authenticate to an Azure Service. </param>
+        /// <param name="endpoint"> server parameter. </param>
+        /// <param name="options"> The options for configuring the client. </param>
+        public AccountsClient(TokenCredential credential, Uri endpoint = null, PurviewAccountClientOptions options = null)
```
The shape of the generated constructor here is wrong.  It should be:

```csharp
AccountsClient(Uri endpoint, TokenCredential credential, PurviewAccountClientOptions options = null)
```

I think some changes to the x-ms-parameterized-host in the swagger will cause M4 to interpret this differently and fix the generation.  You can look at either Scanning or Catalog to see what we did there.

Longer term, we should figure out if we can make the defaults better so that the code generator does the right thing instead of requiring a very specific x-ms-parameterized-host value.

ellismg commented on sdk/purview/Azure.Analytics.Purview.Account/tests/AccountsClientTest.cs:
```diff
@@ -0,0 +1,93 @@
+            Assert.AreEqual("Endpoint=sb://fake_objectId.servicebus.windows.net/;SharedAccessKeyName=AlternateSharedAccessKey;SharedAccessKey=BSDASasdfmasdf123412341234=", listKeyBodyJson.GetProperty("atlasKafkaSecondaryEndpoint"));
+        }
+
+        private static BinaryData GetContentFromResponse(Response r)
```
The linked issue has been fixed, can you try to remove this?

ellismg reviewed (changes requested):
Thanks Crystal!

I left a few comments based on some things I saw that we should go fix.  We'll also want to regenerate the code once the operation ID naming changes I had for the swagger have been merged.

Great to see some tests coming along too!  I did notice the use of default azure credential was causing some issues in CI, my general understanding was didn't use DefaultAzureCredential directly in the tests and instead pulled the credential off the test environment and that got initialized by shared infrastructure.

chunyu3 committed:
resolve comments

chunyu3 committed:
re-generate purveiw account to latest version

chunyu3 committed:
update collection test

chunyu3 committed:
update Readme

chunyu3 committed:
update purview Account API

chunyu3 committed:
record test case

chunyu3 committed:
Merge branch 'main' of https://github.com/Azure/azure-sdk-for-net into feature/llc-purview

chunyu3 committed:
update recorded test case

ellismg committed:
[Purview] Renames + Adopt "Clientlet" Design

- Make some very general client names a little more specific by
  prefixing them with `Purview`.

- Adopt "Clientlet" design, while we work on teaching the code
  generator how to do this.  There are a few changes here:

  - The list methods for Resource Rule Sets and Collections are
    promoted to the root account client.

  - The `collectionName` parameter is promoted to be a client
    parameter instead of an operation parameter.

  - The public constructors for the clientlet types are supressed.

  - `Get<ClientletName>Client()` methods are added to the root client.

ellismg committed:
[Purview] Update Accounts README.md

ellismg committed:
[Purview] Mark methods virtual

ellismg committed:
[Purview] Run Export-API.ps1

ellismg reviewed (approved):
Thanks, Crystal!

This all looks really great!

I pushed a few small changes for naming of the root client types and some manual customizations to adopt the "clientlet" pattern that we want to teach the code generator how to emit.

I feel good about us releasing the fist preview of this library now.
